app/controllers/application_controller.rb

```diff
@@ -72,6 +73,36 @@ def check_your_privilege(name, post = nil, render_error = true)
 
   private
 
+  def stop_the_awful_troll
+    # There shouldn't be any trolls in
+    # the test environment... :D
+    return true if Rails.env.test?
+
+    # Only stop trolls doing things, not looking
+    # at them.
+    return true if request.method.downcase == 'GET'
+
+    # Trolls can't be awful without user accounts.
+    # System is already checking for auth cases.
+    return true if current_user.nil?
+
+    ip = current_user.extract_ip_from(request)
+    email_domain = current_user.email.split('@')[-1]
+
+    is_ip_blocked = BlockedItem.ips.where(value: ip).any?
+    is_mail_blocked = BlockedItem.emails.where(value: current_user.email).any?
+    is_mail_host_blocked = BlockedItem.email_hosts.where(value: email_domain).any?
```

Comment:
Given that this runs against _every_ non-GET request, can we reduce this down to one query?

```suggestion
    ip_block = BlockedItem.active.where(item_type: 'ip', value: ip)
    mail_block = BlockedItem.active.where(item_type: 'email', value: current_user.email)
    mail_host_block = BlockedItem.active.where(item_type: 'email_host', value: email_domain)
    is_blocked = ip_block.or(mail_block).or(mail_host_block)
```

should result in one query with OR clauses.